experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/toEditTree.ts

```diff
@@ -0,0 +1,166 @@
+	context: TreeDataContext,
+	typeSet: TreeTypeSet,
+): CursorAdapter<TNode> {
+	return {
+		value: (node: TNode) => {
+			return isPrimitiveValue(node) || node === null || isFluidHandle(node)
+				? (node as Value)
+				: undefined;
+		},
+		type: (node: TNode) => {
+			if (node === undefined) {
+				throw new Error("undefined node");
+			}
+			if (node === null) {
+				return leaf.null.name;
+			}
+
+			if (isFluidHandle(node)) {
+				return leaf.handle.name;
+			}
+
+			switch (typeof node) {
+				case "number":
+					return leaf.number.name;
+				case "string":
+					return leaf.string.name;
+				case "boolean":
+					return leaf.boolean.name;
+				default: {
+					const possibleTypes = getPossibleTypes(
+						context,
+						typeSet,
+						node as ContextuallyTypedNodeData,
+					);
+					assert(
+						possibleTypes.length !== 0,
+						"data is incompatible with all types allowed by the schema",
+					);
+					assert(
+						possibleTypes.length === 1,
+						"data is compatible with more than one type allowed by the schema",
+					);
+					return possibleTypes[0];
+				}
+			}
+		},
+		keysFromNode: (node: TNode) => {
+			switch (typeof node) {
+				case "object":
+					if (node === null) {
+						return [];
+					} else if (Array.isArray(node)) {
+						return node.length === 0 ? [] : [EmptyKey];
+					} else if (node instanceof Map) {
+						const unfilteredKeys = Array.from(node.keys()) as FieldKey[];
+						// Map proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the entry.
+						const filteredKeys = unfilteredKeys.filter(
+							(key) => node.get(key) !== undefined,
+						);
+						return filteredKeys;
+					} else {
+						// Assume record-like object
+						const objectedNode = node as Record<FieldKey, unknown>;
+
+						const unfilteredKeys = Object.keys(node) as FieldKey[];
+						// Object proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the key/value pair.
+						const filteredKeys = unfilteredKeys.filter((key) => {
+							const value = objectedNode[key];
+							return value !== undefined;
+						});
+						return filteredKeys;
+					}
+				default:
+					return [];
+			}
+		},
+		getFieldFromNode: (node: TNode, key: FieldKey): readonly TNode[] => {
+			// Object.prototype.hasOwnProperty can return true for strings (ex: with key "0"), so we have to filter them out.
+			// Rather than just special casing strings, we can handle them with an early return for all primitives.
+			if (typeof node !== "object") {
+				return [];
+			}
+
+			if (node === null) {
+				return [];
+			}
+
+			if (Array.isArray(node)) {
+				// TODO: verify this
+				return key === EmptyKey || key === rootFieldKey ? node : [];
```

Comment:
@noencke @CraigMacomber does this look right to you? The json cursor implementation only looks at `EmptyKey`, but we have at least some tests with an array as the root field, and they were failing without us also accepting the `rootFieldKey`.

Reply:
the simple tree API does not allow sequence fields in any place other than as the EmptyKey field of a list node. Thus it does not support sequences at the root. Infact it doesn't really support non-required roots very well as currently there is no way to edit or really access the root field.

Reply:
Also, I don't think using Array.isArray here is a good idea. The input here is supposed to be able to be a proxy node tree thing right? List nodes return false for Array.isArray right?

Instead shouldn't you check the schema of the node, and see if its a list schema?